rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/TreeView.java

```diff
@@ -271,150 +270,127 @@ private void enableItems(MenuItem[] items) {
-                    /* Keep host */
-                    if (newNs.getHosts().containsKey(oldHost.getKey())) {
-
-                        Host newHost = newNs.getHosts().get(oldHost.getKey());
-                        for (Entry<String, Node> oldNode : oldHost.getValue().getNodes().entrySet()) {
-                            /* Keep node */
-                            if (newHost.getNodes().containsKey(oldNode.getKey())) {
-                            }
-                            /* Node to be removed */
-                            else if (curNodes.containsKey(oldNode.getKey())) {
-                                this.tree.remove(curNodes.remove(oldNode.getKey()));
-                            }
-                        }
+    }
 
-                    }
-                    /* Host to be removed */
-                    else if (curNodes.containsKey(oldHost.getValue().getId())) {
-                        this.tree.remove(curNodes.remove(oldHost.getValue().getId()));
-                    }
+    private void processNodeSources(List<NodeSource> nodeSources) {
+        if (!nodeSources.isEmpty()) {
+            for (NodeSource nodeSource : nodeSources) {
+                if (nodeSource.isRemoved()) {
+                    removeNodeSource(nodeSource);
+                } else {
+                    addNodeSourceIfNotExists(nodeSource);
+                    changeNodeSourceStatusIfChanged(nodeSource);
                 }
-
-            }
-            /* NodeSource to be removed */
-            else if (curNodes.containsKey(oldNs.getKey())) {
-                this.tree.remove(curNodes.remove(oldNs.getKey()));
             }
         }
+    }
+
+    private void changeNodeSourceStatusIfChanged(NodeSource nodeSource) {
+        if (nodeSource.isChanged()) {
+            TNS curTreeNodeSource = (TNS) currentNodes.get(nodeSource.getSourceName());
+            curTreeNodeSource.rmNS = nodeSource;
+            curTreeNodeSource.setName(getNodeSourceDisplayedDescription(nodeSource, nodeSource.getSourceName()));
+            curTreeNodeSource.setIcon(nodeSource.getIcon());
+        }
+    }
 
-        this.oldNodes = nodes;
+    private void addNodeSourceIfNotExists(NodeSource nodeSource) {
+        if (!currentNodes.containsKey(nodeSource.getSourceName())) {
+            TNS nsTreeNode = new TNS(getNodeSourceDisplayedDescription(nodeSource, nodeSource.getSourceName()),
+                                     nodeSource);
+            nsTreeNode.setIcon(nodeSource.getIcon());
+            tree.add(nsTreeNode, this.tree.getRoot());
+            currentNodes.put(nodeSource.getSourceName(), nsTreeNode);
+        }
 
-        this.treeGrid.markForRedraw();
+    }
+
+    private void removeNodeSource(NodeSource nodeSource) {
+        if (currentNodes.containsKey(nodeSource.getSourceName())) {
+            tree.remove(currentNodes.remove(nodeSource.getSourceName()));
+            currentNodes.remove(nodeSource.getSourceName());
+        }
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either remove or fill this block of code. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00108)